File: plugins/plugin-elizacloud/python/elizaos_plugin_elizacloud/cloud_providers/credit_balance.py
"""
creditBalanceProvider — Credit balance in agent state (60s cache).
"""

from __future__ import annotations

import logging
import time
from typing import TypedDict

from elizaos_plugin_elizacloud.services.cloud_auth_service import CloudAuthService

logger = logging.getLogger("elizacloud.providers.credits")


class ProviderResult(TypedDict, total=False):
    text: str
    values: dict[str, object]


_cache: dict[str, float] | None = None
_cache_at: float = 0.0
_TTL = 60.0


def _format_balance(balance: float) -> ProviderResult:
    low = balance < 2.0
    critical = balance < 0.5
    text = f"ElizaCloud credits: ${balance:.2f}"
    if critical:
        text += " (CRITICAL)"
    elif low:
        text += " (LOW)"
    return ProviderResult(
        text=text,
        values={
            "cloudCredits": balance,
            "cloudCreditsLow": low,
            "cloudCreditsCritical": critical,
        },
    )


async def get_credit_balance(
    auth: CloudAuthService | None = None,
) -> ProviderResult:
    """Get ElizaCloud credit balance with 60s caching."""
    global _cache, _cache_at

    if not auth or not auth.is_authenticated():
        return ProviderResult(text="")

    now = time.time()
    if _cache is not None and (now - _cache_at) < _TTL:
        return _format_balance(_cache["balance"])

    resp = await auth.get_client().get("/credits/balance")
    raw_data = resp.get("data", {})
    if not isinstance(raw_data, dict):
        raw_data = {}
    balance = float(raw_data.get("balance", 0))

    _cache = {"balance": balance}
    _cache_at = now

    if balance < 1.0:
        logger.warning("[CloudCredits] Low balance: $%.2f", balance)

    return _format_balance(balance)


credit_balance_provider: dict[str, object] = {
    "name": "elizacloud_credits",
    "description": "ElizaCloud credit balance",
    "dynamic": True,
    "position": 91,
    "get": get_credit_balance,
}

File: plugins/plugin-elizacloud/typescript/cloud-providers/credit-balance.ts
/**
 * creditBalanceProvider — Credit balance in agent state (60s cache).
 */

import type {
  IAgentRuntime,
  Memory,
  Provider,
  ProviderResult,
  State,
} from "@elizaos/core";
import { logger } from "@elizaos/core";
import type { CloudAuthService } from "../services/cloud-auth";
import type { CreditBalanceResponse } from "../types/cloud";

let cache: { value: number; at: number } | null = null;
const TTL = 60_000;

export const creditBalanceProvider: Provider = {
  name: "elizacloud_credits",
  description: "ElizaCloud credit balance",
  dynamic: true,
  position: 91,

  async get(
    runtime: IAgentRuntime,
    _message: Memory,
    _state: State,
  ): Promise<ProviderResult> {
    const auth = runtime.getService("CLOUD_AUTH") as
      | CloudAuthService
      | undefined;
    if (!auth?.isAuthenticated()) return { text: "" };

    if (cache && Date.now() - cache.at < TTL) return format(cache.value);

    const { data } = await auth
      .getClient()
      .get<CreditBalanceResponse>("/credits/balance");
    cache = { value: data.balance, at: Date.now() };

    if (data.balance < 1.0)
      logger.warn(`[CloudCredits] Low balance: $${data.balance.toFixed(2)}`);
    return format(data.balance);
  },
};

function format(balance: number): ProviderResult {
  const low = balance < 2.0;
  const critical = balance < 0.5;
  let text = `ElizaCloud credits: $${balance.toFixed(2)}`;
  if (critical) text += " (CRITICAL)";
  else if (low) text += " (LOW)";
  return {
    text,
    values: {
      cloudCredits: balance,
      cloudCreditsLow: low,
      cloudCreditsCritical: critical,
    },
  };
}
